cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/RemoteTaskExecutor.java

```diff
@@ -140,13 +142,32 @@ public byte[] runTask(RunnableTaskRequest runnableTaskRequest) throws Exception
             String.format("Received exception %s for %s", e.getMessage(), runnableTaskRequest.getClassName()));
         }
       }, retryStrategy, Retries.DEFAULT_PREDICATE);
+    } catch (ServiceException se) {
+      Exception ex = getTaskException(se);
+      //emit metrics with failed result
+      emitMetrics(startTime, false, runnableTaskRequest, getAttempts(ex));
+      throw ex;
     } catch (Exception e) {
       //emit metrics with failed result
       emitMetrics(startTime, false, runnableTaskRequest, getAttempts(e));
       throw e;
     }
   }
 
+  private Exception getTaskException(ServiceException e) {
+    if (e.getJsonDetails() == null) {
+      // This is not an application-level exception, might be a timeout or similar failure
+      return e;
+    }
+    try {
+      BasicThrowable basicThrowable = GSON.fromJson(e.getJsonDetails(), BasicThrowable.class);
+      return RemoteExecutionException.fromBasicThrowable(basicThrowable);
+    } catch (JsonSyntaxException jse) {
+      return new IllegalStateException(String.format("Service call failed with error code %d", e.getStatusCode()),
+                                       jse);
```

Comment:
just do `e.addSuppresed(jse); throw e;`. This way we won't need to propagate `getStatusCode` from the HTTP layer. We should not bring HTTP here as it's not this layer business if we are using rest or something else (e.g. RPC)

Reply:
fixed